Gem/Code/Source/UserSettings/MultiplayerSampleUserSettings.cpp

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <Atom/RPI.Public/Image/ImageSystem.h>
+#include <Atom/RPI.Public/Image/StreamingImage.h>
+#include <Atom/RPI.Public/Image/StreamingImagePool.h>
+
+#include <AzCore/IO/GenericStreams.h>
+#include <AzCore/Settings/SettingsRegistry.h>
+#include <AzCore/Settings/SettingsRegistryMergeUtils.h>
+#include <AzCore/Utils/Utils.h>
+#include <AzFramework/FileFunc/FileFunc.h>
+#include <IAudioSystem.h>
+#include <UserSettings/MultiplayerSampleUserSettings.h>
+
+namespace MultiplayerSample
+{
+	MultiplayerSampleUserSettings::MultiplayerSampleUserSettings()
+        : m_graphicsApiKey(BaseRegistryKey + FixedString("/ApiName"))
+        , m_masterVolumeKey(BaseRegistryKey + FixedString("/MasterVolume"))
+        , m_textureQualityKey(BaseRegistryKey + FixedString("/TextureQuality"))
+
+	{
+        MultiplayerSampleUserSettingsRequestBus::Handler::BusConnect();
+
+        // Create a full path including filename for the user settings file.
+        m_userSettingsPath = AZ::Utils::GetProjectUserPath();
+        m_userSettingsPath /= "Registry";
+        m_userSettingsPath /= "MultiplayerSampleUserSettings.setreg";
+
+        // Load all of our settings keys, create default values if they don't exist and initialize the engine settings as appropriate.
+        Load();
+    }
+
+	MultiplayerSampleUserSettings::~MultiplayerSampleUserSettings()
+	{
+        MultiplayerSampleUserSettingsRequestBus::Handler::BusDisconnect();
+
+        // Always auto-save the user settings on destruction.
+        Save();
+	}
+
+    void MultiplayerSampleUserSettings::Load()
+    {
+        if (auto* registry = AZ::SettingsRegistry::Get(); registry != nullptr)
+        {
+            AZ::Outcome<AZStd::string, AZStd::string> userSettings = 
+                AzFramework::FileFunc::GetCfgFileContents(AZStd::string(m_userSettingsPath.FixedMaxPathString()));
+
+            if (userSettings.IsSuccess())
+            {
+                // Merge the user settings file under the base "/O3DE/MultiplayerSample/User/Settings" key.
+                // This will ensure that it cannot overwrite any other engine settings.
+                [[maybe_unused]] auto mergeSuccess = registry->MergeSettings(userSettings.GetValue(),
+                    AZ::SettingsRegistryInterface::Format::JsonMergePatch, BaseRegistryKey);
+
+                AZ_Error("UserSettings", mergeSuccess, "Failed to merge user settings into the O3DE registry.");
+            }
```

Comment:
There is a SettingsRegistryMergeUtils function called [MergeSettingsToRegistry_ConfigFile](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Settings/SettingsRegistryMergeUtils.h#L189-L200), that can merge a config file directly to the Settings Registry
```suggestion
            AZ::SettingsRegistryMergeUtils::ConfigParser configParser;
            configParser.m_registryRootPointerPath = BaseRegistryKey;
            [[maybe_unused]] auto mergeSuccess = AZ::SettingsRegistryMergeUtils::MergeSettingsToRegistry_ConfigFile(*registry, m_userSettingsPath.Native(), configParser);
            AZ_Error("UserSettings", mergeSuccess, "Failed to merge user settings into the O3DE registry.");
```

Reply:
Actually this code looks strange.
Why is a Windows style INI/.cfg file being merged to the Settings Registry?

The MergeSettingsFile function can be used directly.
```suggestion
            // Merge the user settings file under the base "/O3DE/MultiplayerSample/User/Settings" key.
            // This will ensure that it cannot overwrite any other engine settings.
            [[maybe_unused]] auto mergeOutcome = registry->MergeSettingsFile(m_userSettingsPath.Native(),
                AZ::SettingsRegistryInterface::Format::JsonMergePatch, BaseRegistryKey);

            AZ_Error("UserSettings", mergeSuccess, "Failed to merge user settings into the O3DE registry with error %s.", mergeOutcome.GetError().c_str());
```

Reply:
I'm not sure that the ConfigParser would work correctly, because this isn't a config file? I admittedly wasn't keen on using GetCfgFileContents() because of the name, but it didn't seem right to rewrite that code here either. I just need the contents of the .setreg file read into memory instead of letting the Settings Registry try to read it because the Settings Registry is using the default FileIOBase, which attempts to read from a PAK file. I needed code that would always read it from a loose file.

I'll add some comments above the GetCfgFileContents() to explain it further.

Reply:
The [AZ::IO::SystemFile::Length and AZ::IO::SystemFile::Read](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/IO/SystemFile.h#L116-L119) function should be used here to read the file then.

There is no need to try to parse the contents of the file at all.

Reply:
I can change it if you feel strongly (though I think I also need to add a trailing \0 after reading, right? and add a ')' to the Length call), but how is putting that implementation inline here any better than using this?

```
        AZ::Outcome<AZStd::string, AZStd::string> GetCfgFileContents(const AZStd::string& filePath)
        {
            if (!AZ::IO::SystemFile::Exists(filePath.c_str()))
            {
                return AZ::Failure(AZStd::string::format("Cfg file '%s' does not exist.", filePath.c_str()));
            }

            // Load the settings file into a string
            AZ::IO::SystemFile settingsFile;
            if (!settingsFile.Open(filePath.c_str(), AZ::IO::SystemFile::SF_OPEN_READ_ONLY))
            {
                return AZ::Failure(AZStd::string::format("Error reading cfg file '%s'.", filePath.c_str()));
            }

            AZStd::string settingsFileContents(settingsFile.Length(), '\0');
            settingsFile.Read(settingsFileContents.size(), settingsFileContents.data());
            settingsFile.Close();
            return AZ::Success(settingsFileContents);

        }
```

It doesn't seem substantially different to me, just more lines of code for me to add to do things that a function in the engine already provides. The only part that's admittedly questionable with my current approach is that "GetCfgFileContents" is a little bit misleading since it isn't a "cfg" file, it's a "setreg" file. But I tried to clarify that in the comments, and maybe this particular method should get renamed in the engine since it's technically capable of reading _any_ text file into memory.

Reply:
I thought that the GetCfgFileContents function was running a regular expression over each line to parse for CFG content, but I got it confused with the [GetValueForKeyInCfgFileContents](https://github.com/o3de/o3de/blob/development/Code/Framework/AzFramework/AzFramework/FileFunc/FileFunc.cpp#L143-L162) function.

Reply:
Side Note: For the https://en.cppreference.com/w/cpp/string/basic_string/resize_and_overwrite function, the null-terminating character is taking into account.
The functor that is specified must the size of the data written to the buffer, which is used to null-terminate by the `resize_and_overwrite_function` as the offset to write the [null-terminator and update the string size](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/std/string/string.h#L1052-L1060)

Reply:
ohhhh, neat! I've never used resize_and_overwrite before, I didn't realize it handled the null terminator, that's pretty cool!